---
title: "MHcut modes: 1 flank vs 2 flanks"
output:
  md_document:
    variant: markdown_github
---

```{r include=FALSE}
knitr::opts_chunk$set(echo=FALSE, message=FALSE, warning=FALSE, fig.width=10, tidy=TRUE)
```

# MHcut modes: 1 flank vs 2 flanks

The first version of MHcut looks for homology in both possible flanks (inner-outer and outer-inner) and picks the ones with highest homology before continuing.
However we might be missing candidates on the other flanks. 
Maybe the flanks with best homology don't have a valid PAM that we could use but the others, although with weaker homology, do.

I ran MHcut in both modes: 1 flank (i.e. best flank only) and 2 flanks.
Now I want to see how many candidates we gained.

```{r load}
## For data manipulation
library(dplyr)
library(magrittr)

## For graphs and tables
library(ggplot2)
library(knitr)
library(RColorBrewer)
palset = c(brewer.pal(9, 'Set1'), brewer.pal(8, 'Set2'))

## Functions
winsor <- function(x, u=10){
  if(any(x>u, na.rm=TRUE)) x[which(x>u)] = u
  x
}
cutn <- function(x){
  cut(x, c(0,1,10,100,1000,Inf))
}

## Import MHcut output
prefix1 = '../clinvar-grch38-pathogenic-deletion'
var1.df = read.table(paste0(prefix1,'-variants.tsv'), as.is=TRUE, header=TRUE, sep='\t')
gui1.df = read.table(paste0(prefix1,'-guides.tsv'), as.is=TRUE, header=TRUE, sep='\t')
var1.df %<>% mutate(size.class=cut(varL, c(0,3,5,10,30,Inf), labels=c('1-3','4-5','6-10','11-30','>30')))

prefix2 = '../clinvar-grch38-pathogenic-deletion-2flanks'
var2.df = read.table(paste0(prefix2,'-variants.tsv'), as.is=TRUE, header=TRUE, sep='\t')
gui2.df = read.table(paste0(prefix2,'-guides.tsv'), as.is=TRUE, header=TRUE, sep='\t')
var2.df %<>% mutate(size.class=cut(varL, c(0,3,5,10,30,Inf), labels=c('1-3','4-5','6-10','11-30','>30')))
```

First how many MH were found in each run:

```{r}
tibble(mode=c('1 flank', '2 flanks'), MH=c(nrow(var1.df), nrow(var2.df))) %>% kable
```

Wow, only 4 variants had MH in both flanks...that's disappointing.

Using the PAM status based on `pamMot`, `pamUniq` and `guidesNoOT` columns, I will count:

- the number of variants with at least one PAM motif.
- the number of variants with at least one PAM motif and unique guide
- the number of variants with at least one PAM motif, a unique guide and no off-target MH.

```{r}
var1.sum = var1.df %>% group_by(Chromosome, Start, Stop) %>% summarize(pam=any(pamMot>0), pam.uniq=any(pamUniq>0), pam.uniq.noOt=any(guidesNoOT>0, na.rm=TRUE)) %>% ungroup %>% select(pam, pam.uniq, pam.uniq.noOt) %>%  summarize_all(sum) %>% mutate(mode='1 flank')
var2.sum = var2.df %>% group_by(Chromosome, Start, Stop) %>% summarize(pam=any(pamMot>0), pam.uniq=any(pamUniq>0), pam.uniq.noOt=any(guidesNoOT>0, na.rm=TRUE)) %>% ungroup %>% select(pam, pam.uniq, pam.uniq.noOt) %>%  summarize_all(sum) %>% mutate(mode='2 flanks')

rbind(var1.sum, var2.sum) %>% select(mode, pam, pam.uniq, pam.uniq.noOt) %>% kable
```

In the end we gained nothing, not surprising considering the number of additional variants that the "2 flanks" mode gives.


## Sanity check: the "1 flank" run is a subset of the "2 flanks" run

```{r}
var2.best.sum = var2.df %>% group_by(Chromosome, Start, Stop) %>% arrange(desc(flankScore), desc(flank)) %>% do(head(., 1)) %>% group_by(Chromosome, Start, Stop) %>% summarize(pam=any(pamMot>0), pam.uniq=any(pamUniq>0), pam.uniq.noOt=any(guidesNoOT>0, na.rm=TRUE)) %>% ungroup %>% select(pam, pam.uniq, pam.uniq.noOt) %>%  summarize_all(sum) %>% mutate(mode='2 flanks')

rbind(var1.sum, var2.best.sum) %>% select(mode, pam, pam.uniq, pam.uniq.noOt) %>% kable
```

What !? 

Hum, could it be that some variants from the "2 flanks" mode are not included in the other run?

```{r}
var1.s = var1.df %>% group_by(Chromosome, Start, Stop) %>% summarize(pam=any(pamMot>0), pam.uniq=any(pamUniq>0), pam.uniq.noOt=any(guidesNoOT>0, na.rm=TRUE))
var2.s = var2.df %>% group_by(Chromosome, Start, Stop) %>% summarize(pam=any(pamMot>0), pam.uniq=any(pamUniq>0), pam.uniq.noOt=any(guidesNoOT>0, na.rm=TRUE))
var.s = merge(var1.s, var2.s, by=c('Chromosome', 'Start', 'Stop'), all=TRUE, suffixes=c('.1','.2'))
var.s %>% filter(is.na(pam.1)) %>% kable
```

Ah, these two variants were not included in the "1 flank" mode results.

```{r}
var.s %>% filter(is.na(pam.1)) %>% select(Chromosome, Start, Stop) %>% merge(var2.df) %>% kable
```

One is a large deletion, the other is more surprising. 
I checked the cartoon for the second one and it's because the best pair of flanks were filtered out right away, most likely by the `m1L>=3` default filter.
We might want to first check for these criteria and then chose the pair of flanks. 
Or change the scoring system that we use to choose a pair of flanks.
