---
title: "Summary of MHcut output"
output:
  md_document:
    variant: markdown_github
---

```{r include=FALSE}
knitr::opts_chunk$set(echo=FALSE, message=FALSE, warning=FALSE, fig.width=10, tidy=TRUE)
```

# Summary of MHcut output

This report shows some simple analysis to quickly check the output of MHcut.
The goal is to get an idea of the distribution of each metric in the output and check if something is off.

```{r load}
## For data manipulation
library(dplyr)
library(magrittr)

## For graphs and tables
library(ggplot2)
library(knitr)
library(RColorBrewer)
palset = c(brewer.pal(9, 'Set1'), brewer.pal(8, 'Set2'))

## Functions
winsor <- function(x, u=10){
  if(any(x>u, na.rm=TRUE)) x[which(x>u)] = u
  x
}
cutn <- function(x){
  cut(x, c(0,1,10,100,1000,Inf))
}

## Import MHcut output
prefix = '../clinvar-grch38-all-deletion-1bp'
var.df = read.table(paste0(prefix,'-variants.tsv'), as.is=TRUE, header=TRUE, sep='\t')
gui.df = read.table(paste0(prefix,'-guides.tsv'), as.is=TRUE, header=TRUE, sep='\t')
var.df %<>% mutate(size.class=cut(varL, c(0,3,5,10,30,Inf), labels=c('1-3','4-5','6-10','11-30','>30')))
```

## At the variant level

### Extremes

Longest MH:

```{r extmh}
var.df %>% arrange(desc(mhL)) %>% select(Chromosome, Start, Stop, mhL, mh1L) %>% head %>% kable
```

Most distant MHs:

```{r extmhdist}
var.df %>% arrange(desc(mhDist)) %>% select(Chromosome, Start, Stop, mhL, mh1L, varL, mhDist) %>% head %>% kable
```


### Variant length vs MH length

```{r var}
ggplot(var.df, aes(x=winsor(varL, u=20))) + geom_histogram(binwidth=1) + theme_bw()
ggplot(var.df, aes(x=winsor(mhL, u=20))) + geom_histogram(binwidth=1) + theme_bw()
var.df %>% mutate(varL=winsor(varL, 50), mhL=winsor(mhL, 50)) %>% group_by(mhL, varL) %>% summarize(n=cutn(n())) %>% ggplot(aes(x=varL, y=mhL, fill=n)) + geom_tile() + theme_bw() + ggtitle('Winsorized at 50')
ggplot(var.df, aes(x=winsor(mh1L, u=20))) + geom_histogram(binwidth=1) + theme_bw()
```

### Homology

```{r hom}
ggplot(var.df, aes(x=hom)) + geom_histogram() + theme_bw()
var.df %>% mutate(mhL=winsor(mhL,80)) %>% group_by(mhL, hom) %>% summarize(n=cutn(n())) %>% ggplot(aes(x=mhL, y=hom, fill=n)) + geom_tile() + theme_bw()
```

### Distance between MHs

```{r mhdist}
var.df %>% ggplot(aes(x=mhDist)) + geom_histogram() + theme_bw() + facet_wrap(~size.class, scales='free')
var.df %>% mutate(varL=winsor(varL, 50), mhDist=winsor(mhDist, 50)) %>% group_by(size.class, varL, mhDist) %>% summarize(n=cutn(n())) %>% ggplot(aes(x=varL, y=mhDist, fill=n)) + geom_tile() + theme_bw() + geom_abline(linetype=2) + ggtitle('Winsorized at 50')
```


### PAM status

Defining the PAM status based on `pamMot`, `pamUniq` and `guidesNoOT` columns.

*The `NA`s represents variants that were not tested for off-target because too large.*

```{r pam}
var.df %<>% mutate(pam.status=ifelse(guidesNoOT>0, 'no off-target',  'off-targets'), pam.status=ifelse(pamUniq==0, 'no unique PAM', pam.status), pam.status=ifelse(pamMot==0, 'no PAM', pam.status))
ggplot(var.df, aes(x=pamMot, fill=pam.status)) + geom_histogram(binwidth=1) + theme_bw() + facet_grid(size.class~., scales='free')
ggplot(var.df, aes(x=pamUniq, fill=pam.status)) + geom_histogram(binwidth=1) + theme_bw() + facet_grid(size.class~., scales='free')
var.df %>% filter(pamUniq>0) %>% ggplot(aes(x=guidesNoOT, fill=pam.status)) + geom_histogram(binwidth=1) + theme_bw() + facet_grid(size.class~., scales='free') + ggtitle('PAM available')
var.df %>% filter(pamUniq>0, guidesNoOT==0) %>% ggplot(aes(x=guidesMinOT, fill=pam.status)) + geom_histogram(binwidth=1) + theme_bw() + facet_grid(size.class~., scales='free') + ggtitle('PAM available, off-target MH')
```

## At the guide level

```{r guidespam}
gui.df %<>% mutate(size.class=cut(varL, c(0,3,5,10,30,Inf), labels=c('1-3','4-5','6-10','11-30','>30')), pam.status=ifelse(guidesNoOT>0, 'no off-target',  'off-targets'), pam.status=ifelse(pamUniq==0, 'no unique PAM', pam.status), pam.status=ifelse(pamMot==0, 'no PAM', pam.status))
ggplot(gui.df, aes(x=pamMot, fill=pam.status)) + geom_histogram(binwidth=1) + theme_bw() + facet_grid(size.class~., scales='free')
ggplot(gui.df, aes(x=botScore, fill=pam.status)) + geom_histogram() + theme_bw() + facet_grid(size.class~., scales='free')
ggplot(gui.df, aes(x=botSize, fill=pam.status)) + geom_histogram(binwidth=1) + theme_bw() + facet_grid(size.class~., scales='free')
```

### Best off-target MH

For now the *best off-target* MH is based on the MMEJ score.

```{r best}
gui.df %>% mutate(varL=winsor(varL, 100), botVarL=winsor(botVarL, 100)) %>% group_by(varL, botVarL) %>% summarize(n=cutn(n())) %>% ggplot(aes(x=varL, y=botVarL, fill=n)) + geom_tile() + theme_bw() + geom_abline(linetype=2) + ggtitle('Winsorized at 100')
```

Best off-target is the longest off-target ?

```{r botsize}
gui.df %>% group_by(largestOffTgt, botSize) %>% summarize(n=cutn(n())) %>% ggplot(aes(x=largestOffTgt, y=botSize, fill=n)) + theme_bw() + geom_abline(linetype=2) + geom_tile()
```

## Conclusions

The output looks fine to me.


